### PR TITLE
Implement Knight Tour algorithm using Backtracking.

### PR DESCRIPTION
**Pull request by TruongNhanNguyen**
## Description

This PR implements the Knight's Tour algorithm in Rust, along with unit tests and documentation updates.

## Type of change

- [x] New feature (non-breaking change which adds functionality) :rocket: The purpose of this PR is to add the Knight's Tour algorithm to the project, which solves the classic chess problem of finding a sequence of knight moves that visits every square of a chessboard exactly once. By adding this algorithm, we enhance the project's functionality and provide a useful feature for users who are interested in chess algorithms.

## Checklist :heavy_check_mark: 

- [x] I ran bellow commands using the latest version of **rust nightly**.
- [x] I ran `cargo clippy --all -- -D warnings` just before my last commit and fixed any issue that was found.
- [x] I ran `cargo fmt` just before my last commit.
- [x] I ran `cargo test` just before my last commit and all tests passed.
- [x] I added my algorithm to the corresponding `mod.rs` file within its own folder, and in any parent folder(s).
- [x] I added my algorithm to `DIRECTORY.md` with the correct link.
- [x] I checked `COUNTRIBUTING.md` and my code follows its guidelines.

## Additional Notes :pencil2: 

This PR includes comprehensive unit tests for the Knight's Tour algorithm, covering various edge cases and scenarios. Additionally, the algorithm is well-documented to help users understand its functionality and usage.

### CODE REVIEW

File: src/backtracking/knight_tour.rs [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+            vec![14, 9, 20, 5, 16],
+            vec![19, 2, 7, 22, 11],
+            vec![8, 13, 24, 17, 4],
+            vec![25, 18, 3, 12, 23],
+        ];
+
+        // Compare each cell of the generated tour matrix with the expected matrix
+        for i in 0..5 {
+            assert_eq!(tour_matrix[i], expected_matrix[i]);
+        }
```
Comment (vil02) [inline src/backtracking/knight_tour.rs:143]: What do you think about making these tests more _direst_:
```suggestion
        assert_eq!(tour_matrix, expected_matrix)
```

File: src/backtracking/knight_tour.rs [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+            vec![17, 2, 33, 8, 13, 10],
+            vec![32, 35, 24, 21, 28, 5],
+            vec![23, 18, 3, 30, 9, 20],
+            vec![36, 31, 22, 19, 4, 29],
+        ];
+
+        // Compare each cell of the generated tour matrix with the expected matrix
+        for i in 0..6 {
+            assert_eq!(tour_matrix[i], expected_matrix[i]);
+        }
```
Comment (vil02) [inline src/backtracking/knight_tour.rs:165]:
```suggestion
        assert_eq!(tour_matrix, expected_matrix)
```

File: src/backtracking/knight_tour.rs [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+            vec![43, 58, 3, 50, 41, 24, 7, 20],
+            vec![48, 51, 46, 55, 26, 21, 12, 15],
+            vec![57, 44, 53, 4, 23, 14, 25, 6],
+            vec![52, 47, 56, 45, 54, 5, 22, 13],
+        ];
+
+        // Compare each cell of the generated tour matrix with the expected matrix
+        for i in 0..8 {
+            assert_eq!(tour_matrix[i], expected_matrix[i]);
+        }
```
Comment (vil02) [inline src/backtracking/knight_tour.rs:189]:
```suggestion
        assert_eq!(tour_matrix, expected_matrix)
```

File: src/backtracking/knight_tour.rs [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    /// Finds the Knight's Tour starting from the specified position.
+    ///
+    /// # Arguments
+    ///
+    /// * `start_x` - The x-coordinate of the starting position.
+    /// * `start_y` - The y-coordinate of the starting position.
+    ///
+    /// # Returns
+    ///
+    /// A tuple containing a boolean indicating whether a tour was found and the tour matrix if found.
```
Comment (vil02) [inline src/backtracking/knight_tour.rs:102]: Please explain what is the _tour matrix_.

File: src/backtracking/knight_tour.rs [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+        if !self.solve_tour(start_x as i32, start_y as i32, 1) {
+            println!("No solution exists");
+            return (false, vec![]);
+        }
+
+        println!("Knight's Tour:");
+        (true, self.board.clone())
+    }
+}
```
Comment (vil02) [inline src/backtracking/knight_tour.rs:119]: What do you thing about having just one public function? Its signature should look something like that

```rust
pub fn find_tour(size_x: usize, size_y: usize, start_x: usize, start_y: usize) -> (bool, Vec<Vec<usize>>)
```

In my opinion nothing else should be available/visible for the external user.

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/backtracking/knight_tour.rs (hunk 1 above)
Comment (TruongNhanNguyen) [reply src/backtracking/knight_tour.rs:143]: Well, It's will make the code more flexible and readable, especially if dealing with larger matrices. Done! :rocket:

File: src/backtracking/knight_tour.rs (hunk 3 above)
Comment (TruongNhanNguyen) [reply src/backtracking/knight_tour.rs]: Done :rocket:

File: src/backtracking/knight_tour.rs (hunk 2 above)
Comment (TruongNhanNguyen) [reply src/backtracking/knight_tour.rs]: Done :rocket:

File: src/backtracking/knight_tour.rs (hunk 4 above)
Comment (TruongNhanNguyen) [reply src/backtracking/knight_tour.rs:102]: In my cpde, the term _tour matrix_ refers to a matrix representing the solution to the Knight's Tour problem. The Knight's Tour problem is about finding a sequence of moves for a knight on a chessboard such that the knight visits every square exactly once.

In the context of my implementation:

- The `KnightTour` struct represents the implementation of the Knight's Tour problem.
- The `board` field within the `KnightTour` struct is a two-dimensional vector (`Vec<Vec<i32>>`) representing the chessboard. Each element of this matrix holds an integer value representing the order in which the knight visits that particular square. A value of `0` indicates that the square has not been visited yet.
- The `find_tour` method attempts to find a solution to the Knight's Tour problem starting from a specified position on the chessboard. If a solution is found, it returns a tuple containing a Boolean indicating success (`true`) and the tour matrix.
- The tour matrix returned by `find_tour` is essentially the `board` field of the `KnightTour` struct. It represents the sequence of moves made by the knight on the chessboard, with each cell containing the order in which the knight visited that square.

So, the tour matrix is essentially a record of the knight's path across the chessboard, providing a visual representation of the solution to the problem.

File: src/backtracking/knight_tour.rs (hunk 5 above)
Comment (TruongNhanNguyen) [reply src/backtracking/knight_tour.rs:119]: - Yeah, I think it is a reasonable design choice, especially if we want to encapsulate the implementation details of Knight's Tour and expose only necessary functionalities for user.
- By having a single public function, we can hide the internal details of the `KnightTour` struct and its methods, providing a simpler and cleaner interface for users.
- By providing the necessary parameters `(size_x, size_y, start_x, start_y)` for the chessboard size and starting position, users have the flexibility to customize their Knight's Tour problem without needing to understand the internal workings of the implementation.

Comment (TruongNhanNguyen) [reply src/backtracking/knight_tour.rs:119]: I will apply your suggestion in the next changes (commit).

Comment (TruongNhanNguyen) [reply src/backtracking/knight_tour.rs:119]: I will change the name of the public function from `find_tour` to `find_knight_tour` to make the function more descriptive.

File: src/backtracking/knight_tour.rs (hunk 4 above)
Comment (TruongNhanNguyen) [reply src/backtracking/knight_tour.rs:126]: Do I need to add explanation for _tour matrix_ to the comment.

Comment (vil02) [reply src/backtracking/knight_tour.rs:126]: Yes!

File: src/backtracking/knight_tour.rs [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+///
+/// # Returns
+///
+/// A tuple containing a boolean indicating whether a tour was found and the tour matrix if found.
+pub fn find_knight_tour(
+    size_x: usize,
+    size_y: usize,
+    start_x: usize,
+    start_y: usize,
+) -> (bool, Vec<Vec<usize>>) {
```
Comment (vil02) [inline src/backtracking/knight_tour.rs:22]:
```suggestion
) -> Option(Vec<Vec<usize>>) {
```

When the solution does not exist one can return `None`.

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/backtracking/knight_tour.rs (hunk 4 above)
Comment (TruongNhanNguyen) [reply src/backtracking/knight_tour.rs:126]: Done.

File: src/backtracking/knight_tour.rs (hunk 6 above)
Comment (TruongNhanNguyen) [reply src/backtracking/knight_tour.rs]: Done.

File: src/backtracking/knight_tour.rs [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+/// Represents the KnightTour struct which implements the Knight's Tour problem.
+struct KnightTour {
+    board_size_x: usize,
+    board_size_y: usize,
+    board: Vec<Vec<usize>>,
+    moves: [(i32, i32); 8],
+}
+
+impl KnightTour {
+    /// Constructs a new KnightTour instance with the given board size.
```
Comment (vil02) [inline src/backtracking/knight_tour.rs:39]:
```suggestion
    const MOVES: [(i32, i32); 8] = [
        (2, 1),
        (1, 2),
        (-1, 2),
        (-2, 1),
        (-2, -1),
        (-1, -2),
        (1, -2),
        (2, -1),
    ];
    /// Constructs a new KnightTour instance with the given board size.
```

File: src/backtracking/knight_tour.rs [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+    ///
+    /// A boolean indicating whether a solution was found.
+    fn solve_tour(&mut self, x: i32, y: i32, move_count: i32) -> bool {
+        if move_count == (self.board_size_x * self.board_size_y) as i32 {
+            return true;
+        }
+
+        for i in 0..8 {
+            let next_x = x + self.moves[i].0;
+            let next_y = y + self.moves[i].1;
```
Comment (vil02) [inline src/backtracking/knight_tour.rs:104]:
```suggestion
            let next_x = x + KnightTour::MOVES[i].0;
            let next_y = y + KnightTour::MOVES[i].1;
```

File: src/backtracking/knight_tour.rs [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+        }
+
+        self.board[start_x][start_y] = 1;
+
+        if !self.solve_tour(start_x as i32, start_y as i32, 1) {
+            println!("No solution exists");
+            return None;
+        }
+
+        println!("Knight's Tour:");
```
Comment (vil02) [inline src/backtracking/knight_tour.rs:143]:
```suggestion
            return None;
        }

        self.board[start_x][start_y] = 1;

        if !self.solve_tour(start_x as i32, start_y as i32, 1) {
            return None;
        }

```

Comment (TruongNhanNguyen) [reply src/backtracking/knight_tour.rs:143]: Yeah, I have removed all unnecessary `println` statements.

File: src/backtracking/knight_tour.rs (hunk 7 above)
Comment (TruongNhanNguyen) [reply src/backtracking/knight_tour.rs:39]: Done :rocket:

File: src/backtracking/knight_tour.rs (hunk 8 above)
Comment (TruongNhanNguyen) [reply src/backtracking/knight_tour.rs:104]: Done.

File: src/backtracking/knight_tour.rs [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+    let mut tour = KnightTour::new(size_x, size_y);
+    tour.find_tour(start_x, start_y)
+}
+
+/// Represents the KnightTour struct which implements the Knight's Tour problem.
+struct KnightTour {
+    board_size_x: usize,
+    board_size_y: usize,
+    board: Vec<Vec<usize>>,
+    moves: [(i32, i32); 8],
```
[the remark this answers is not in the document]
Comment (TruongNhanNguyen) [reply src/backtracking/knight_tour.rs:35]: Done

File: src/backtracking/knight_tour.rs [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+        // Define the expected tour matrix for an 5x5 board
+        let tour_matrix = find_knight_tour(5, 5, 2, 1);
+        assert_eq!(tour_matrix, None);
+    }
+
+    #[test]
+    fn test_invalid_start_position() {
+        let tour_matrix = find_knight_tour(8, 8, 10, 10);
+        assert_eq!(tour_matrix, None);
+    }
```
[the remark this answers is not in the document]
Comment (TruongNhanNguyen) [reply src/backtracking/knight_tour.rs:218]: Well, organize test cases in this way make them more clear and compact. But I will change the name of tests from `tc_5x5` to `test_knight_tour5x5` (similar to `tc_6x6` and `tc_8x8` and another tests) to make them more descriptive.
```suggestion
    macro_rules! test_find_knight_tour {
        ($($name:ident: $tc:expr,)*) => {
        $(
            #[test]
            fn $name() {
                let (size_x, size_y, start_x, start_y, expected) = $tc;
                assert_eq!(find_knight_tour(size_x, size_y, start_x, start_y), expected);
            }
        )*
        }
    }
    test_find_knight_tour! {
        test_knight_tour_5x5: (5, 5, 0, 0, Some(vec![
            vec![1, 6, 15, 10, 21],
            vec![14, 9, 20, 5, 16],
            vec![19, 2, 7, 22, 11],
            vec![8, 13, 24, 17, 4],
            vec![25, 18, 3, 12, 23],
        ])),
        test_knight_tour_6x6: (6, 6, 0, 0, Some(vec![
            vec![1, 16, 7, 26, 11, 14],
            vec![34, 25, 12, 15, 6, 27],
            vec![17, 2, 33, 8, 13, 10],
            vec![32, 35, 24, 21, 28, 5],
            vec![23, 18, 3, 30, 9, 20],
            vec![36, 31, 22, 19, 4, 29],
        ])),
        test_knight_tour_8x8: (8, 8, 0, 0, Some(vec![
            vec![1, 60, 39, 34, 31, 18, 9, 64],
            vec![38, 35, 32, 61, 10, 63, 30, 17],
            vec![59, 2, 37, 40, 33, 28, 19, 8],
            vec![36, 49, 42, 27, 62, 11, 16, 29],
            vec![43, 58, 3, 50, 41, 24, 7, 20],
            vec![48, 51, 46, 55, 26, 21, 12, 15],
            vec![57, 44, 53, 4, 23, 14, 25, 6],
            vec![52, 47, 56, 45, 54, 5, 22, 13],
        ])),
        test_no_solution: (5, 5, 2, 1, None),
        test_invalid_start_position: (8, 8, 10, 10, None),
    }
```

File: src/backtracking/knight_tour.rs [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+    ///
+    /// A boolean indicating whether a solution was found.
+    fn solve_tour(&mut self, x: i32, y: i32, move_count: i32) -> bool {
+        if move_count == (self.board_size_x * self.board_size_y) as i32 {
+            return true;
+        }
+
+        for i in 0..8 {
+            let next_x = x + KnightTour::MOVES[i].0;
+            let next_y = y + KnightTour::MOVES[i].1;
```
Comment (vil02) [inline src/backtracking/knight_tour.rs:103]:
```suggestion
        for &(dx, dy) in KnightTour::MOVES.iter() {
            let next_x = x + dx;
            let next_y = y + dy;
```

File: src/backtracking/knight_tour.rs [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+    /// # Arguments
+    ///
+    /// * `x` - The current x-coordinate of the knight.
+    /// * `y` - The current y-coordinate of the knight.
+    /// * `move_count` - The current move count.
+    ///
+    /// # Returns
+    ///
+    /// A boolean indicating whether a solution was found.
+    fn solve_tour(&mut self, x: i32, y: i32, move_count: i32) -> bool {
```
Comment (vil02) [inline src/backtracking/knight_tour.rs:96]: is the a reason that `move_count` is of the type `i32`? Would it be possible to make it some _unsigned_ type?

File: src/backtracking/knight_tour.rs [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+    start_y: usize,
+) -> Option<Vec<Vec<usize>>> {
+    let mut tour = KnightTour::new(size_x, size_y);
+    tour.find_tour(start_x, start_y)
+}
+
+/// Represents the KnightTour struct which implements the Knight's Tour problem.
+struct KnightTour {
+    board_size_x: usize,
+    board_size_y: usize,
```
Comment (vil02) [inline src/backtracking/knight_tour.rs:33]: These two fields are not needed. Both sizes could be queried using the `board`.

**Review by vil02 — CHANGES_REQUESTED** [review]
Regarding the failing clippy check, please _rebase_ yout branch (cf. #700).

File: src/backtracking/knight_tour.rs (hunk 12 above)
Comment (TruongNhanNguyen) [reply src/backtracking/knight_tour.rs:103]: Done.

File: src/backtracking/knight_tour.rs (hunk 13 above)
Comment (TruongNhanNguyen) [reply src/backtracking/knight_tour.rs:96]: Yes, I have changed `move_count` type from `i32` to `u32`

File: src/backtracking/knight_tour.rs (hunk 14 above)
Comment (TruongNhanNguyen) [reply src/backtracking/knight_tour.rs]: Solved.

**Comment by TruongNhanNguyen** [comment]
Let's has a final look @vil02

File: src/backtracking/knight_tour.rs [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+    /// # Arguments
+    ///
+    /// * `x` - The current x-coordinate of the knight.
+    /// * `y` - The current y-coordinate of the knight.
+    /// * `move_count` - The current move count.
+    ///
+    /// # Returns
+    ///
+    /// A boolean indicating whether a solution was found.
+    fn solve_tour(&mut self, x: i32, y: i32, move_count: u32) -> bool {
```
Comment (vil02) [inline src/backtracking/knight_tour.rs:91]: `move_count` should be of the type `usize`. Also `x` and `y` should be of the type [`isize`](https://doc.rust-lang.org/std/primitive.isize.html#).

File: src/backtracking/knight_tour.rs [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+mod tests {
+    use super::*;
+
+    macro_rules! test_find_knight_tour {
+        ($($name:ident: $tc:expr,)*) => {
+        $(
+            #[test]
+            fn $name() {
+                let (size_x, size_y, start_x, start_y, expected) = $tc;
+                assert_eq!(find_knight_tour(size_x, size_y, start_x, start_y), expected);
```
Comment (vil02) [inline src/backtracking/knight_tour.rs:148]: What do you think about adding an extra check for the test data?
```suggestion
                if expected.is_some() {
                    assert_eq!(expected.clone().unwrap()[start_x][start_y], 1)
                }
                assert_eq!(find_knight_tour(size_x, size_y, start_x, start_y), expected);
```

**Review by vil02 — CHANGES_REQUESTED** [review]
Almost done. Please _sort out_ the types: `i32` to `isize` etc.

File: src/backtracking/knight_tour.rs [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+    ///
+    /// * `x` - The current x-coordinate of the knight.
+    /// * `y` - The current y-coordinate of the knight.
+    /// * `move_count` - The current move count.
+    ///
+    /// # Returns
+    ///
+    /// A boolean indicating whether a solution was found.
+    fn solve_tour(&mut self, x: i32, y: i32, move_count: u32) -> bool {
+        if move_count == (self.board.len() * self.board[0].len()) as u32 {
```
[the remark this answers is not in the document]
Comment (TruongNhanNguyen) [reply src/backtracking/knight_tour.rs:92]: These two methods make the code more encapsulated (OOP). I added them into the implementation. Appreciate your suggestions.

File: src/backtracking/knight_tour.rs (hunk 15 above)
Comment (TruongNhanNguyen) [reply src/backtracking/knight_tour.rs:91]: It's a good refactoring, changes are clarity and consistency.

File: src/backtracking/knight_tour.rs (hunk 16 above)
Comment (TruongNhanNguyen) [reply src/backtracking/knight_tour.rs:148]: In my point of view, adding an extra check to verify that the starting position in the expected result is indeed marked as 1 is a good idea. It adds an additional level of validation to ensure that the tour starts from the correct position as specified in the test case. So this can help catch potential issues such as incorrect initialization or indexing errors.

File: src/backtracking/knight_tour.rs [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+            vec![38, 35, 32, 61, 10, 63, 30, 17],
+            vec![59, 2, 37, 40, 33, 28, 19, 8],
+            vec![36, 49, 42, 27, 62, 11, 16, 29],
+            vec![43, 58, 3, 50, 41, 24, 7, 20],
+            vec![48, 51, 46, 55, 26, 21, 12, 15],
+            vec![57, 44, 53, 4, 23, 14, 25, 6],
+            vec![52, 47, 56, 45, 54, 5, 22, 13],
+        ])),
+        test_no_solution: (5, 5, 2, 1, None),
+        test_invalid_start_position: (8, 8, 10, 10, None),
```
[the remark this answers is not in the document]
Comment (TruongNhanNguyen) [reply src/backtracking/knight_tour.rs:180]: Done.

**Review by vil02 — APPROVED** [review]
